### ISSUE TITLE
[CUSTOMER] updateflag.awk fail with timeout or be blocked with "Retrying flag update"  when provisioning a bunch of node 

### ISSUE THREAD

**Issue by immarvin** [xcat2/xcat-core#2325]
the is reported from the customer,  when they reprovision 516 nodes concurrently, some nodes delayed a significant time to update its status, and several nodes failed to update their status to "booted". in the xcat.log, they observed this:
````
updateflag.awk: Retrying flag update
````

**Comment by immarvin** [comment]
fixed in https://github.com/xcat2/xcat-core/pull/2326

**Comment by immarvin** [comment]
another PR related to this issue https://github.com/xcat2/xcat-core/pull/2318

**Comment by Obihoernchen** [comment]
Is it feasible (without causing other problems) to cherry-pick both PRs for 2.12.3?

**Comment by immarvin** [comment]
hi @Obihoernchen , the PR https://github.com/xcat2/xcat-core/pull/2326 have some problems during review.
It is ok to cherry-pick the PR https://github.com/xcat2/xcat-core/pull/2318 for 2.12.3

**Comment by Obihoernchen** [comment]
I did some more research on this because it was way too annoying to have those issues every boot.

First of all I've patched xcatd of all our 2.12.3 service xcat nodes with https://github.com/xcat2/xcat-core/pull/2318/commits/3d3f367f69d18e09e849101d7110e16536cc2030

Now I don't see the `xcatd: possible BUG encountered by xCAT DB worker SIGPIPE xcatd: [...]` error anymore.
But it doesn't seem to fix anything else except this log message.

------------------------------

If you boot more than 5 nodes simultaneously, 50% of them will fail to boot.
The named daemon will go to (and stay at) 100% cpu load on the corresponding xcat service node.
Actually named gets ddosed by all compute nodes with boot issues.

The reason for this is the `/tmp/updateflag` script in `initrd-stateless.gz`.
The script is created during the genimage process: https://github.com/xcat2/xcat-core/blob/master/xCAT-server/share/xcat/netboot/sles/genimage#L1751
​
The problem with this script is, that it has no loop exit condition and no retry delay at all.
So if xcat is unable to update the nodestatus it will run the following code indefinitely: 
```awk
        ns = "/inet/tcp/0/" xcatdhost "/" xcatiport
        print "xCAT_xcatd" |& ns
        while(1) {
                ns |& getline
                if(\$0 == "ready")
                        print ARGV[3] |& ns
                if(\$0 == "done")
                        break
        }

```
This results in kind of ddos against the service node DNS because it resolves the hostname every time.
On top of that the node will not continue to boot, if it is unable to update the nodestatus.

Because I didn't want to rebuild the whole image with a custom genimage script, I just edited initrd-stateless.gz and replaced /tmp/updateflag with the following (based on https://github.com/xcat2/xcat-core/blob/master/xCAT/postscripts/updateflag.awk):
```awk
#!/usr/bin/awk -f
BEGIN {
    xcatdhost = ARGV[1]
    xcatdport = ARGV[2]
    flag = ARGV[3]
    if (!flag) flag = "next"
    ns = "/inet/tcp/0/" xcatdhost "/" xcatdport
    loop = 0
    while(1) {
        if((ns |& getline) > 0)
            print $0
        else {
            print "Retrying flag update"
            print "updateflag.awk: Retrying flag update"
            close(ns)
            system("sleep 10")
            loop = loop + 1
        }
        if($0 == "ready")
            print flag |& ns
        if($0 == "done")
            break
        if(loop > 10) {
            print "flag update failed"
            print "updateflag.awk: flag update failed"
            break
        }
    }
    close(ns)
    exit 0
}
```
Now at least all nodes boot fine again and the service node named doesn't go crazy anymore, despite the missing nodestatus update:

```
[Thu Jan  5 15:56:49 2017]udevd[388]: specified group 'floppy' unknown
[Thu Jan  5 15:56:49 2017]
[Thu Jan  5 15:56:49 2017]udevd[388]: specified group 'cdrom' unknown
[Thu Jan  5 15:56:49 2017]
[Thu Jan  5 15:56:49 2017]udevd[388]: specified group 'tape' unknown
[Thu Jan  5 15:56:49 2017]
err, eth0: fopen `/var/run/ntp/servers.eth0': No such file or directory
          inet addr:10.4.4.62  Bcast:10.4.7.255  Mask:255.255.252.0
Retrying flag update
updateflag.awk: Retrying flag update
Retrying flag update
updateflag.awk: Retrying flag update
Retrying flag update
updateflag.awk: Retrying flag update
Retrying flag update
updateflag.awk: Retrying flag update
Retrying flag update
updateflag.awk: Retrying flag update
Retrying flag update
updateflag.awk: Retrying flag update
Retrying flag update
updateflag.awk: Retrying flag update
Retrying flag update
updateflag.awk: Retrying flag update
Retrying flag update
updateflag.awk: Retrying flag update
Retrying flag update
updateflag.awk: Retrying flag update
Retrying flag update
updateflag.awk: Retrying flag update
flag update failed
updateflag.awk: flag update failed
Getting http://i20xcat2cl//install/netboot/sles11.4/x86_64/compute/rootimg.cpio.gz...
--2017-01-05 14:58:58--  http://i20xcat2cl//install/netboot/sles11.4/x86_64/compute/rootimg.cpio.gz
Resolving i20xcat2cl... 10.4.4.2
Connecting to i20xcat2cl|10.4.4.2|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1205751747 (1.1G) [application/x-gzip]
Saving to: `rootimg.cpio.gz'
80% [==============================>        ] 966,983,969 20.2M/s  eta 17s     
```

But the nodestatus is wrong. Somehow xcat is not able to update the status.
Once xcat hits this issue, even booting one node doesn't update the status anymore.
The mysql DB is no problem. It's doing almost nothing the whole time and there is not a single slow query. 

```
$ nodels i20r01c02s02,i20r01c02s04,i20r01c02s06,i20r01c02s08,i20r01c02s10,i20r01c02s12,i20r01c05s02,i20r01c05s04,i20r01c05s06,i20r01c05s08,i20r01c05s10,i20r01c05s12,i20r01c06s02,i20r01c06s04,i20r01c06s06,i20r01c06s08,i20r01c06s10,i20r01c06s12 nodelist.status
i20r01c02s02: powering-on
i20r01c02s04: powering-on
i20r01c02s06: powering-on
i20r01c02s08: powering-on
i20r01c02s10: powering-on
i20r01c02s12: powering-on
i20r01c05s02: powering-on
i20r01c05s04: powering-on
i20r01c05s06: powering-on
i20r01c05s08: powering-on
i20r01c05s10: powering-on
i20r01c05s12: powering-on
i20r01c06s02: powering-on
i20r01c06s04: powering-on
i20r01c06s06: powering-on
i20r01c06s08: powering-on
i20r01c06s10: powering-on
i20r01c06s12: powering-on

$ xdsh i20r01c02s02,i20r01c02s04,i20r01c02s06,i20r01c02s08,i20r01c02s10,i20r01c02s12,i20r01c05s02,i20r01c05s04,i20r01c05s06,i20r01c05s08,i20r01c05s10,i20r01c05s12,i20r01c06s02,i20r01c06s04,i20r01c06s06,i20r01c06s08,i20r01c06s10,i20r01c06s12 uptime
i20r01c06s12:  16:58pm  up   1:02,  0 users,  load average: 0.10, 0.06, 0.14
i20r01c02s06:  16:58pm  up   1:02,  0 users,  load average: 0.10, 0.04, 0.13
i20r01c05s04:  16:58pm  up   1:02,  0 users,  load average: 0.06, 0.03, 0.12
i20r01c02s12:  16:58pm  up   1:02,  0 users,  load average: 0.07, 0.05, 0.12
i20r01c02s02:  16:58pm  up   1:02,  0 users,  load average: 0.00, 0.02, 0.12
i20r01c02s08:  16:58pm  up   1:02,  0 users,  load average: 0.12, 0.07, 0.14
i20r01c06s04:  16:58pm  up   1:02,  0 users,  load average: 0.07, 0.04, 0.13
i20r01c06s10:  16:58pm  up   1:02,  0 users,  load average: 0.00, 0.01, 0.11
i20r01c02s10:  16:58pm  up   1:02,  0 users,  load average: 0.00, 0.01, 0.12
i20r01c05s02:  16:58pm  up   1:01,  0 users,  load average: 0.00, 0.03, 0.13
i20r01c05s06:  16:58pm  up   1:02,  0 users,  load average: 0.07, 0.05, 0.13
i20r01c05s08:  16:58pm  up   1:02,  0 users,  load average: 0.00, 0.01, 0.12
i20r01c05s10:  16:58pm  up   1:01,  0 users,  load average: 0.00, 0.04, 0.13
i20r01c02s04:  16:58pm  up   1:02,  0 users,  load average: 0.01, 0.04, 0.12
i20r01c05s12:  16:58pm  up   1:02,  0 users,  load average: 0.01, 0.03, 0.12
i20r01c06s02:  16:58pm  up   1:02,  0 users,  load average: 0.00, 0.01, 0.11
i20r01c06s06:  16:58pm  up   1:02,  0 users,  load average: 0.00, 0.01, 0.12
i20r01c06s08:  16:58pm  up   1:02,  0 users,  load average: 0.01, 0.03, 0.13
```

Having 10 different updateflag.awk scripts in xcat source code is not that good I think.
All these scripts are in a different state (there are different versions for genesis, netboot, osimage etc).
Some of them are strings inside another script, some of them are files. But all are different (with or without sleep, loop abort, ...)
A unified updateflag.awk script should help to prevent issues like this I think.

**Comment by whowutwut** [comment]
@Obihoernchen  I just cherry-picked commit 3d3f367f69d18e09e849101d7110e16536cc2030 to the 2.12 branch.  (#2318) The snapshot builds will be created for 2.12 which is at version 2.12.4 version but there's currently no plan to release a 2.12.4 version at this time.
